File: 5_percent_py/03_largest_prime_factor.py
import _common_functions as cf

def largest_prime_factor(number):
  """
  number: number on which is tested factors (prime or not)
  return: list of factors of number and the max factor of that list
  """
  factors = []
  for elt in cf.list_factors(number):
    if cf.is_prime(elt):
      factors.append(elt)
  return max(factors)

# 1st attempt: i solved this one using pure mathematics since
# i couldn't make an efficient enough program to
# find the correct answer without waiting ten years
# 2nd attempt: used my list_factors algorithm, the problem is now solved in 0.0812s
assert largest_prime_factor(600851475143) == 6857, "largest_prime_factor(600851475143), 6857"
print(cf.average_running_time(10, largest_prime_factor, 600851475143))

File: 5_percent_c/03_largest_prime_factor.c
#include <stdio.h>
#include <math.h>

int isPrime(long long number, long long big_number)
{
	if (big_number % number == 0)
	{
		for(int i = 2; i<number; i++)
		{
			if (number % i == 0)
				return 0;
		}
		return 1;
	}
	return 0;
}

int main()
{
	long long big_number = 600851475143;
	long long factor = 775147;

	while(!(isPrime(factor, big_number)))
		factor--;
	printf("%lld is the largest prime factor of %lld\n", factor, big_number);
	return 0;	
}
